openwisp_controller/config/base/config.py

```diff
@@ -330,10 +330,29 @@ def manage_vpn_clients(cls, action, instance, pk_set, **kwargs):
         else:
             templates = pk_set
 
-        # Delete VPN clients that are not associated with current templates
-        instance.vpnclient_set.exclude(
-            template_id__in=instance.templates.values_list("id", flat=True)
-        ).delete()
+        if len(pk_set) != templates.filter(required=True).count():
```

Comment:
This checks that all the applied templates in `pk` set are not required templates. If they are, then we don't delete the VpnClient in this step. The comment below explains the rest.